### PR TITLE
Add prompt learning for T5

### PR DESCRIPTION
**Pull request by HeyyyyyyG**
# What does this PR do ?

Implement prompt learning for T5 models.

**Collection**: [NLP]

# Changelog 
- A base prompt learning dataset class (will support both T5 and GPT in the future).
- A T5-specific prompt learning dataset class.
- A base prompt learning model class (will support both T5 and GPT in the future).
- A T5-specific prompt learning model class.
- Scripts for training and inference.
- Config files for training and inference. 
- Temporarily fix a bug in T5 decode function to support prompt learning inference.

# Usage

```
python megatron_t5_prompt_learning.py --config-name=megatron_t5_prompt_learning.yaml
python megatron_t5_prompt_learning_eval.py t5_model_file=models/megatron_t5_220m_f16.nemo virtual_prompt_model_file=models/p_tuning_sent_t5.nemo
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Review by vadam5 — CHANGES_REQUESTED** [review]
Over all good work! I left a few comments. In addition, could you also remove all old ptuning files? This should at a minimum include these 3:

```
nlp/models/language_modeling/megatron_ptune_t5_model.py
nlp/data/glue_benchmark/t5_ptune_dataset.py
nlp/modules/common/t5_prompt_encoder.py
```

File: nemo/collections/nlp/data/language_modeling/megatron/t5_prompt_learning_dataset.py
Comment (HeyyyyyyG) [reply nemo/collections/nlp/data/language_modeling/megatron/t5_prompt_learning_dataset.py:100]: I think it is necessary. The decoder input should be `<bos> A B` when the label is `A B C`.

File: nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py
Comment (HeyyyyyyG) [reply nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py:67]: I don't see how we can do that. I have put those initializations which are not dependent on the frozen model in the base model's init class. Other initializations such as the `tokenizer`, `hidden_size`, and `psuedo_token_ids` are dependent on the frozen model. They can only be initialized after we load the frozen model.

File: nemo/collections/nlp/models/language_modeling/megatron_base_prompt_learning_model.py
Comment (MaximumEntropy) [inline nemo/collections/nlp/models/language_modeling/megatron_base_prompt_learning_model.py:67]: You should be able to get this from `frozen_model.autocast_dtype` instead of this if-else.

File: nemo/collections/nlp/models/language_modeling/megatron_base_prompt_learning_model.py
Comment (MaximumEntropy) [inline nemo/collections/nlp/models/language_modeling/megatron_base_prompt_learning_model.py:116]: Why are we getting both `autocast_dtype` and `float_type`? They should always be the same no?

Comment (MaximumEntropy) [reply nemo/collections/nlp/models/language_modeling/megatron_base_prompt_learning_model.py:116]: Is there a case where this can be different?

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py
Comment (MaximumEntropy) [inline nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:445]: This bit is being refactored in this PR - https://github.com/NVIDIA/NeMo/pull/4210. Do we have to add all these additional args to make it work?

File: nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py
Comment (MaximumEntropy) [inline nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py:88]: You shouldn't have to do any autocasting if precision is configured properly for the prompt learning and the frozen model. If there is a precision mismatch between pre-training and prompt-learning, you should override the precision in the frozen model before restoring. The user should only configure `trainer.precision` from which the prompt-learning model and frozen model should obtain float types.

File: nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py
Comment (MaximumEntropy) [inline nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py:128]: This is where you want to override precision. Set `t5_cfg.precision = trainer.precision`.

File: nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py
Comment (MaximumEntropy) [inline nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py:102]: You shouldn't call the parent model's forward function since this will not work with pipeline parallel and will need significant refactoring to work with it. Please write another method that uses apex's `forward_backward_pipelining_without_interleaving` or `forward_backward_no_pipelining`.

File: nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py
Comment (MaximumEntropy) [inline nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py:288]: I think there is an issue on the main branch right now wherein `encoder_input` will just be ignored even if you pass it. I think @michalivne is working on a fix. Lets communicate with him about this.

File: nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py
Comment (MaximumEntropy) [inline nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py:347]: If your data parallel size is > 1 at predict time, this will only compute accuracy per data parallel rank. Since you use the DistributedSampler, this will only be the accuracy on a subset of your dataset. You can either use `TorchMetrics` like I do in the t5 finetuning class or do an allgather object from all data parallel ranks like we do in NMT (see https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/nlp/models/machine_translation/megatron_nmt_model.py#L337)

**Review by MaximumEntropy — CHANGES_REQUESTED** [review]
A bunch of comments: Maybe better communicated over a call. Lets try and chat about this next week :)

**Comment by MaximumEntropy** [comment]
Thanks for the great PR! Excited to use this for a bunch of models I'm training :)

Just wanted to add a general comment. When writing some of the initial fine-tuning class and filenames, I used "T5" everywhere and I actually meant to say "encoder-decoder". Most of what you implemented is agnostic to the pre-training approach and should work pretty much any encoder-decoder model. Can we try and rename things to make it generic to the fact that is should work with any enc-dec model (except ofcourse your sentinel addition trick).

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py
Comment (HeyyyyyyG) [reply nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:445]: No. I just temporarily modify the args so that `enc_input` can be utilized in the `decode` function. I see the changes in that PR. If it gets merged first, I can change my codes based on it.

File: nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py
Comment (HeyyyyyyG) [reply nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py:288]: Hi @michalivne ! I slightly change the args in `get_forward_output_only_func` so that `encoder_input` can be utilized. I see you are working on a fix. Please let me know when your fix is done and merged.

File: Jenkinsfile
Comment (HeyyyyyyG) [reply Jenkinsfile:2991]: Yes. Only ~10 lines.

File: nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py
Comment (vadam5) [reply nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py:386]: @michalivne wanted to follow up on this. We are shooting to merge in Jiaqi's PR before the code freeze next week. Any update on when your PR will be merged?

**Review by vadam5 — COMMENTED** [review]
Great work! Just a few small comments, mostly on the config file.

File: examples/nlp/language_modeling/megatron_t5_prompt_learning.py
Comment (HeyyyyyyG) [reply examples/nlp/language_modeling/megatron_t5_prompt_learning.py:74]: Yes

**Review by vadam5 — APPROVED** [review]
Great work Jiaqi! Let's just wait for @MaximumEntropy's approval before merging.

**Review by vadam5 — APPROVED** [review]
LGTM!

**Review by MaximumEntropy — APPROVED** [review]
LGTM! Awesome PR :)

**Review by vadam5 — APPROVED** [review]
LGTM!

**Pull request merged**
